app/questionnaire/navigation.py

```diff
@@ -31,110 +31,141 @@ def build_navigation(self, current_group_id, current_group_instance):
-                    logger.debug("building navigation", group_id=group['id'])
-                    if len(self.completed_blocks) > 0 and set(self.routing_path[:-1]).issubset(self.completed_blocks):
-                        navigation.append(self._build_single_navigation(group, current_group_id, first_location))
-                elif last_group_id != group['id'] and not is_summary_or_confirm_group:
-                    logger.debug("building navigation", group_id=group['id'])
-                    navigation.append(self._build_single_navigation(group, current_group_id, first_location))
+            repeating_rule = SchemaHelper.get_repeat_rule(
+                SchemaHelper.get_group(self.survey_json, first_group_id))
+            if repeating_rule:
+                navigation.extend(self._build_repeating_navigation(repeating_rule, section, current_group_id,
+                                                                   current_group_instance, last_block_id))
+            else:
+                navigation.append(self._build_single_navigation(section, current_group_id,
+                                                                Location(first_group_id, 0, first_block_id),
+                                                                last_block_id))
 
         return navigation
 
-    def _should_skip_group(self, current_group_instance, group):
+    def _get_non_skipped_groups(self, section):
+        non_skipped_groups = []
+        for group_id in section['group_order']:
+            if not self._should_skip_group(group_id):
+                non_skipped_groups.append(group_id)
+
+        return non_skipped_groups
 
-        skip_group = False
+    def _should_skip_group(self, group_id):
+        skip_nav = False
+        group = SchemaHelper.get_group(self.survey_json, group_id)
         skip_conditions = SchemaHelper.get_skip_condition(group)
 
         if skip_conditions:
-            skip_group = evaluate_skip_conditions(skip_conditions, self.metadata, self.answer_store,
-                                                  current_group_instance)
-        return skip_group
+            skip_nav = evaluate_skip_conditions(skip_conditions, self.metadata, self.answer_store)
+
+        return skip_nav
+
+    def _get_last_group_and_block_id(self, group_list, group_instance):
+        """
+        Navigate backwards through each non skipped group in a section
+         checking each block in reverse until a non skipped
+         block is obtained.
+
+        :param group_list: List of groups to check
+        :param group_instance: Used for repeating groups
+        :return:
+        """
+        for group_id in reversed(group_list):
+            group = SchemaHelper.get_group(self.survey_json, group_id)
+            for block in reversed(group['blocks']):
+                skip_cons = block.get('skip_conditions')
+                if evaluate_skip_conditions(skip_cons, self.metadata, self.answer_store, group_instance) is False:
```

Comment:
`group_instance` is the current group instance the user is on when `build_navigation` is called, and doesn't change each time `_get_last_group_and_block_id` is called. I think for now we should remove this argument and let `evaluate_skip_conditions` use the default of `0`.  This will mean that for every repeat of a group the completed block will be the same, but we should tackle that problem when we look at using the routing path to determine if a section is complete.